### PR TITLE
CRM-19776: Allow to set component for case types hook

### PR DESCRIPTION
**Pull request by omarabuhussein**
The current problem with created case types via hook_civicrm_caseTypes or hook_civicrm_managed is that it doesn't allow you to specify the component ID for the attached activity types and CiviCase component will be used by default.

This change allow developers to specify the component for managed case types in the XML file , and if it defined , it will be used to specify the component for the managed activities for that case type .. otherwise ; the default CiviCase component will be used .

Example usage :

if an extension define  a new component let's call it "TestComponent" , and the developer want to create a new case type when installing this extension let's call it "TestCase" and there will be some activities attached to this case type but they have to be part of "TestComponent" not "CiviCase" component ..

then a file called at xml/case/TestCase.xml can be created which contain something like this :

``` xml

<?xml version="1.0" encoding="iso-8859-1" ?>

<CaseType>
  <name>TestCase</name>
  <weight>100</weight>
  <componentName>TestComponent</componentName>
  <ActivityTypes>
    <ActivityType>
      <name>Background Check</name>
    </ActivityType>
    <ActivityType>
      <name>References Check</name>
    </ActivityType>
  </ActivityTypes>
</CaseType>
```

in this case the two activities "Background Check" and "References Check" when get created via managed entities handler will have component_id = [TestComponent ID].

but in case "<componentName>TestComponent</componentName>"  is not defined in the xml file , the default "CiviCase" will be used .

---

 * [CRM-19776: CaseTypes hook doesn't allow you to specify the component for child activity types](https://issues.civicrm.org/jira/browse/CRM-19776)

### CODE REVIEW

**Comment by omarabuhussein** [comment]
@totten  I saw that you are the one who mainly worked on managed entities so I want to bother you a  and ask you if you could check this PR whenever you have some time . Thanks

**Comment by totten** [comment]
* The `array_map("unserialize", array_unique(array_map("serialize", $result)))` is a neat trick.
 * If I'm reading correctly, this changes the signature of  `getDeclaredActivityTypes()` and `getAllDeclaredActivityTypes()` (i.e. different `@return` structure). The doc blocks need to match the new return types. IMHO, a good example of a docblock describing nested arrays would be `CRM_Utils_Hook::managed($entities)`. 
   * (You don't need to duplicate the docblock in every function that uses the same array... just describe it once and use `@see` for the others...)
 * What happens when a case-type includes a mix of (a) its own activity types (like "Background Check") and (b) existing/standard activity types (like "Meeting" or "Phone Call"). What values would one expect for the "Meeting"'s `component_id`?

**Comment by totten** [comment]
Also - did you do any testing with what happens you edit/save these case types using `Administer => CiviCase => Case Types`? Off the top of my head, any change in the XML format needs (at least) corresponding updates to `convertDefinitionToXML()` and `convertXmlToDefinition()`. Fixing these should amenable to a test-driven approach (`CRM_Case_BAO_CaseTypeTest::definitionProvider()`).

**Comment by omarabuhussein** [comment]
@totten first thanks for taking a look at it  ... 

> What happens when a case-type includes a mix of (a) its own activity types (like "Background Check") and (b) existing/standard activity types (like "Meeting" or "Phone Call"). What values would one expect for the "Meeting"'s component_id?

if the component ID is not specified , it will use the default (CiviCase) component ID .

> did you do any testing with what happens you edit/save these case types using Administer => CiviCase => Case Types?

I did a new installation via the buildkit and navigated accross the created case types with no problem ( but of course I need to do another test ) but tbh I didn't tried to create a new case from the UI ( as far as I remember ) .


I will fix the points you talked about and do another round of more detailed testing and get back to you with results , thanks so much.

**Comment by omarabuhussein** [comment]
Hey @totten ,

First I think I misunderstood this question :

> What happens when a case-type includes a mix of (a) its own activity types (like "Background Check") and (b) existing/standard activity types (like "Meeting" or "Phone Call"). What values would one expect for the "Meeting"'s component_id?

well... I choose defining the component name for the case type instead for activity types is to make it easier to work with (for example if we have 15 activity type for a case type , then we only define the component once instead of 15 times).
The other thing that I believe it doesn't make sense to have a case types attached with activity types from different components ( the Idea of using component in activity types is to act as a sort of separating activity types into different groups so why a single case type would have activity types from different groups ?!! )  .

now regarding this :

> Also - did you do any testing with what happens you edit/save these case types using Administer => CiviCase => Case Types? Off the top of my head, any change in the XML format needs (at least) corresponding updates to convertDefinitionToXML() and convertXmlToDefinition(). Fixing these should amenable to a test-driven approach (CRM_Case_BAO_CaseTypeTest::definitionProvider()).

First , I tested it and it works fine when adding cases from UI , but regarding changing convertDefinitionToXML() and convertXmlToDefinition() , well .. I thought about it and I think it is not needed ... The first reason is because the definition field is only contain only ( activity types , activity sets and roles ) of the case type and In my patch I set the component for the whole case type and not per activity type inside the case type so the component name is not part of activity type XML , the other thing I believe the only need for component Name is for creation of these activity types but I don't think  it is important to see them in the definition field .


regarding the docblocks I just fixed them and pushed the changes .

**Comment by omarabuhussein** [comment]
@totten what do you think ^ ?

**Comment by guanhuan** [comment]
jenkins, please test this

**Comment by omarabuhussein** [comment]
@totten ...

**Comment by colemanw** [comment]
I may be missing something, but shouldn't you define your activity types via hook_civicrm_managed separately from your case types? I'm pretty sure that hook will allow you to specify component id when defining new activity types.

**Comment by omarabuhussein** [comment]
I recall that I've used hook_civicrm_managed but it was also have the same problem of hook_civicrm_case but will take another look

**Comment by omarabuhussein** [comment]
hmmm , I just read your comment again and I think you are right ( not sure why this didn't came to my mind again ) .. and I also just tested that and it works actually .. but the only problem with that is it will require your to define the activity types twice ( in the case type definition and  separately to create the activities ) as well as you need to ensure that the activities are created before the case  type.

Anyway , I could totally go with that approach but I think this with this fix the solution become more cleaner .

@colemanw

File: CRM/Case/XMLProcessor/Process.php [hunk 1]
```diff
@@ -313,28 +313,40 @@ public function activityTypes($activityTypesXML, $maxInst = FALSE, $isLabel = FA
...
     }
 
     if (!empty($caseTypeXML->ActivitySets) && $caseTypeXML->ActivitySets->ActivitySet) {
       foreach ($caseTypeXML->ActivitySets->ActivitySet as $activitySetXML) {
         if ($activitySetXML->ActivityTypes && $activitySetXML->ActivityTypes->ActivityType) {
           foreach ($activitySetXML->ActivityTypes->ActivityType as $activityTypeXML) {
-            $result[] = (string) $activityTypeXML->name;
+            $result[] = array(
+              'name' => (string) $activityTypeXML->name,
+              'component_name' => $componentName);
```
Comment (colemanw) [inline CRM/Case/XMLProcessor/Process.php]: Closing array parenthesis should go on the next line (also see L333 for the same issue).

**Comment by colemanw** [comment]
This looks ok to me. @totten do you have any concerns about merging it?

**Comment by omarabuhussein** [comment]
fixed

**Comment by omarabuhussein** [comment]
I will close this since @colemanw suggestion fixed what I wanted to achieve and this become unneeded anymore ... thanks guys

**Pull request closed without merging**